src/ImageSharp/Formats/Qoi/QoiDecoderCore.cs

```diff
@@ -0,0 +1,287 @@
+                                };
+                                pixel.FromRgba32(readPixel);
+                                pixelArrayPosition = GetArrayPosition(readPixel);
+                                previouslySeenPixels[pixelArrayPosition] = readPixel;
+                                break;
+
+                            // Get green difference in 6 bits and red and blue differences
+                            // depending on the green one
+                            case QoiChunk.QoiOpLuma:
+                                byte diffGreen = (byte)(operationByte & 0b00111111),
+                                    currentGreen = (byte)((previousPixel.G + (diffGreen - 32)) % 256),
+                                    nextByte = (byte)stream.ReadByte(),
+                                    diffRedDG = (byte)(nextByte >> 4),
+                                    diffBlueDG = (byte)(nextByte & 0b00001111),
+                                    currentRed = (byte)((diffRedDG - 8 + (diffGreen - 32) + previousPixel.R) % 256),
+                                    currentBlue = (byte)((diffBlueDG - 8 + (diffGreen - 32) + previousPixel.B) % 256);
+                                readPixel = previousPixel with { R = currentRed, B = currentBlue, G = currentGreen };
+                                pixel.FromRgba32(readPixel);
+                                pixelArrayPosition = GetArrayPosition(readPixel);
+                                previouslySeenPixels[pixelArrayPosition] = readPixel;
+                                break;
+
+                            // Repeating the previous pixel 1..63 times
+                            case QoiChunk.QoiOpRun:
+                                byte repetitions = (byte)(operationByte & 0b00111111);
+                                if (repetitions is 62 or 63)
+                                {
+                                    ThrowInvalidImageContentException();
+                                }
+
+                                readPixel = previousPixel;
+                                pixel.FromRgba32(readPixel);
+                                for (int k = -1; k < repetitions; k++, j++)
+                                {
+                                    if (j == this.header.Width)
+                                    {
+                                        j = 0;
+                                        i++;
+                                    }
+
+                                    pixels[j, i] = pixel;
+                                }
+
+                                j--;
+                                continue;
+
+                            default:
+                                ThrowInvalidImageContentException();
+                                return;
+                        }
+
+                        break;
+                }
+
+                pixels[j, i] = pixel;
```

Comment:
Avoid use this indexer!
Take `Span<TPixel> row = pixels.DangerousGetRowSpan(i);` in the outer for loop. Then you can do `row[j] = pixel`.

Same for the encoder!

Reply:
Something funny is that initially, when creating both functions, I used the `[j, i]` indexer because I didn't understand `pixels.DangerousGetRowSpan(i);`, but now that I've been so deep into the project, I wonder why I didn't do that before xD

Reply:
Even I do that sometimes.